xm-commons-timeline/src/main/java/com/icthh/xm/commons/timeline/config/DomainEventTimelineInterceptor.java

```diff
@@ -0,0 +1,141 @@
+        }
+        return null;
+    }
+
+    // TODO
+    private boolean isIgnoredRequest(HttpServletRequest request, HttpServletResponse response) {
+        return false;
+    }
+
+    private Map<String, String> getUserDetails(OAuth2Authentication auth) {
+        Map<String, String> details = null;
+        if (auth.getDetails() != null) {
+            details = Map.class.cast(OAuth2AuthenticationDetails.class.cast(auth.getDetails()).getDecodedDetails());
+        }
+        details = firstNonNull(details, new HashMap<>());
+        return details;
+    }
+
+    private void publishEvent(HttpServletRequest request, HttpServletResponse response, String tenant, String userLogin, String userKey) {
+        DomainEvent domainEvent = DomainEvent.builder()
+            .id(UUID.randomUUID())
+            .tenant(tenant)
+            .userKey(userKey)
+            .payload(createPayload(request, response))
+            .build();
+        eventPublisher.publish(DefaultDomainEventSource.HTTP.name(), domainEvent);
+    }
+
+    @SneakyThrows
+    private HttpDomainEventPayload createPayload(HttpServletRequest request, HttpServletResponse response) {
+        ContentCachingResponseWrapper responseCacheWrapperObject = new ContentCachingResponseWrapper(response);
+
+        String requestBody = IOUtils.toString(request.getReader());
+        String responseBody = IOUtils.toString(responseCacheWrapperObject.getContentInputStream(), StandardCharsets.UTF_8);
+
+        HttpDomainEventPayload payload = new HttpDomainEventPayload();
+        payload.setMethod(request.getMethod());
+        payload.setUrl(request.getRequestURI());
+        payload.setQueryString(request.getQueryString());
+        payload.setRequestLength(request.getContentLengthLong());
+        payload.setRequestBody(requestBody);
+        payload.setResponseBody(responseBody);
+        payload.setResponseLength((long) responseBody.length());
+        payload.setRequestHeaders(getRequestHeaders(request));
+        payload.setResponseHeaders(getResponseHeaders(response));
+        payload.setResponseCode(response.getStatus());
+        payload.setExecTime(MdcUtils.getExecTimeMs());
+        return payload;
+    }
+
+    private HttpHeaders getRequestHeaders(HttpServletRequest httpRequest) {
+        Map<String, List<String>> headersMap = Collections.list(httpRequest.getHeaderNames())
+            .stream()
+            .collect(Collectors.toMap(
+                Function.identity(),
+                h -> Collections.list(httpRequest.getHeaders(h))
+            ));
+
+        return HttpHeaders.of(headersMap, (s1, s2) -> true);
+    }
+
+    private HttpHeaders getResponseHeaders(HttpServletResponse httpServletResponse) {
+        Map<String, List<String>> headersMap = httpServletResponse.getHeaderNames().stream()
+            .collect(Collectors.toMap(
+                Function.identity(),
+                h -> new ArrayList<>(httpServletResponse.getHeaders(h))
```

Comment:
why different implementation in same case at line 125?

Reply:
Enumeration<String> getHeaderNames = HttpServletRequest.getHeaderNames()
Enumeration<String> getHeaders = HttpServletRequest.getHeaders("{header-name}")
 
Collection<String> getHeaderNames = HttpServletResponse.getHeaderNames()
Collection<String> getHeaders = HttpServletResponse.getHeaders("{header-name}")